big_tests/tests/mam_helper.hrl

```diff
@@ -18,15 +18,31 @@
 
 -include_lib("exml/include/exml.hrl").
 
+-define(safe_fail(FmtStr, Args),
+        ct:fail("~ts", [lager_format:format(FmtStr, Args, 25000)])).
+
 -define(assert_equal(E, V), (
-    [ct:fail("ASSERT EQUAL~n\tExpected ~p~n\tValue ~p~n", [(E), (V)])
+    [?safe_fail("ASSERT EQUAL~n\tExpected ~p~n\tValue ~p~n", [(E), (V)])
      || (E) =/= (V)]
     )).
 
+
+%% If map is extra, format it as it in "k = v" format.
 -define(assert_equal_extra(E, V, Extra), (
-    [ct:fail("assert_equal_extra(~s, ~s)~n\tExpected ~p~n\tValue ~p~nExtra ~p~n",
-             [(??E), (??V), (E), (V), (Extra)])
-     || (E) =/= (V)]
+    (fun(E0, V0, Extra0) when is_map(Extra0), E0 =/= V0 ->
+            %% preserve map ordering
+            ExtraArgs = maps:fold(fun(MK,MV,A) -> A ++ [MK,MV] end, [], Extra0),
+            ExtraFmt = lists:append(lists:duplicate(maps:size(Extra0), "~n~p = ~p")),
+            ?safe_fail("assert_equal_extra(~s, ~s)~n\t"
+                        "Expected ~p~n\tValue ~p" ++ ExtraFmt,
+                       [??E, ??V, E0, V0|ExtraArgs]);
+        (E0, V0, Extra0) when E0 =/= V0 ->
+            ?safe_fail("assert_equal_extra(~s, ~s)~n\t"
+                        "Expected ~p~n\tValue ~p~nExtra ~p~n",
+                       [??E, ??V, E0, V0, Extra0]);
+        (_, _, _) ->
```

Comment:
Can we name and define this fun somwhere (e.g `fun perserve_ordering/3`) and call it? It's hard to read what does it really do here